src/unix/aio.c

```diff
@@ -0,0 +1,352 @@
+}
+
+sysreturn io_submit(aio_context_t ctx_id, long nr, struct iocb **iocbpp)
+{
+    struct aio *aio;
+    if (!ctx_id) {
+        return -EFAULT;
+    }
+    if (!(aio = aio_from_ring(current->p, ctx_id))) {
+        return -EINVAL;
+    }
+    int io_ops;
+    for (io_ops = 0; io_ops < nr; io_ops++) {
+        sysreturn rv = iocb_enqueue(aio, iocbpp[io_ops]);
+        if (rv) {
+            if (io_ops == 0) {
+                return rv;
+            } else {
+                break;
+            }
+        }
+    }
+    return io_ops;
+}
+
+closure_function(5, 1, sysreturn, io_getevents_bh,
+                 struct aio *, aio, long, min_nr, long, nr, struct io_event *, events, thread, t,
+                 u64, flags)
+{
+    struct aio *aio = bound(aio);
+    struct io_event *events = bound(events);
+    thread t = bound(t);
+    aio_ring ring = aio->ring;
+    aio_lock(aio);
+    unsigned int head = ring->head;
+    unsigned int tail = ring->tail;
+    if (head >= aio->nr) {
+        head = 0;
+    }
+    if (tail >= aio->nr) {
+        tail = 0;
+    }
+    while (head != tail) {
+        runtime_memcpy(&events[aio->copied_evts], &ring->events[head],
+                sizeof(struct io_event));
+        if (++head == aio->nr) {
+            head = 0;
+        }
+        if (++aio->copied_evts == bound(nr)) {
+            break;
+        }
+    }
+    ring->head = head;
+    ring->tail = tail;
+    aio_unlock(aio);
+    sysreturn rv = 0;
+    if (flags & BLOCKQ_ACTION_NULLIFY) {
+        rv = -EINTR;
+    } else if ((aio->copied_evts < bound(min_nr)) &&
+            !(flags & BLOCKQ_ACTION_TIMEDOUT)) {
+        return BLOCKQ_BLOCK_REQUIRED;
+    }
+    aio->bq = 0;
+    if (aio->copied_evts != 0) {
+        rv = aio->copied_evts;
+    }
+    if (flags & BLOCKQ_ACTION_BLOCKED) {
+        thread_wakeup(t);
+    }
+    closure_finish();
+    return set_syscall_return(t, rv);
+}
+
+sysreturn io_getevents(aio_context_t ctx_id, long min_nr, long nr,
+        struct io_event *events, struct timespec *timeout)
+{
+    if (!ctx_id || !events) {
+        return -EFAULT;
+    }
+    struct aio *aio;
+    if ((nr <= 0) || (nr < min_nr) ||
+            !(aio = aio_from_ring(current->p, ctx_id))) {
+        return -EINVAL;
+    }
+    aio->copied_evts = 0;
+    aio->bq = current->thread_bq;
+    return blockq_check_timeout(aio->bq, current,
+            closure(heap_general(aio->kh), io_getevents_bh, aio, min_nr, nr,
+                    events, current), false, CLOCK_ID_MONOTONIC,
+            timeout ? time_from_timespec(timeout) : 0, false);
+}
+
+sysreturn io_destroy(aio_context_t ctx_id)
+{
+    if (!ctx_id) {
+        return -EFAULT;
+    }
+    struct aio *aio;
+    if (!(aio = aio_from_ring(current->p, ctx_id))) {
+        return -EINVAL;
+    }
+
+    /* Wait for any ongoing operations to be completed. */
+    while ((volatile unsigned int) aio->ongoing_ops) {
+        struct io_event event;
+        io_getevents(ctx_id, 1, 1, &event, 0);
```

Comment:
This may not ever return if the check in io_getevents blocks, right? If so this might need to be refactored such that the rest of the destroy is gated on events completion.

Reply:
Right. I changed io_getevents_bh() to take an io_completion as an additional argument, and defined io_destroy_complete() as a closure function that can be instantiated as an io_completion, so that io_destroy() can invoke io_getevents_bh() with this completion.